src/firewall/core/fw_policy.py

```diff
@@ -61,9 +61,11 @@ def get_active_policies_not_derived_from_zone(self):
 
         return active_policies
 
-    def get_policy(self, policy):
-        p = self._fw.check_policy(policy)
-        return self._policies[p]
+    def get_policy(self, policy, required=True):
+        p = self._policies.get(policy, None)
+        if p is None and required:
+            raise FirewallError(errors.INVALID_POLICY, policy)
+        return p
```

Comment:
It's the third iteration, because its repeated for different types.

Let's look at `check_zone()`/`get_zone()`. At  `master` branch we have:

```python
    # src/firewall/core/fw.py
    def check_zone(self, zone):
        _zone = zone
        if not _zone or _zone == "":
            _zone = self.get_default_zone()
        if _zone not in self.zone.get_zones():
            raise FirewallError(errors.INVALID_ZONE, _zone)
        return _zone

    # src/firewall/core/fw_zone.py
    def get_zone(self, zone):
        z = self._fw.check_zone(zone)
        return self._zones[z]   
```

On `master`, `Firewall.check_zone()` would O(n) iterate over all zones. Even `FirewallZone.get_zone()` would call `self._fw.check_zone()` to get the validation from `Firewall.check_zone()`. Note that `FirewallZone._zones` is private API, so `Firewall.check_zone()` can't easily do better, unless it
- accesses `self._fw_zone._zones`, thereby weakening the encapsulation and making responsibilities less clear.
- delegate to `FirewallZone` for the validation.

The branch does the latter and turns it around to:

```python
    # src/firewall/core/fw.py
    def check_zone(self, zone):
        return self.zone.check_zone(zone)

    # src/firewall/core/fw_zone.py
    def check_zone(self, zone):
        return self.get_zone(zone).name
    def get_zone(self, zone, required=True):
        if not zone:
            zone = self._fw.get_default_zone()
        z = self._zones.get(zone, None)
        if z is None and required:
            raise FirewallError(errors.INVALID_ZONE, zone)
        return z
```

Yes, the branch just moves some code around. While no functional change, the benefit is that responsibilities are moved closer together. Naming of things and layout/place of the code matters, to a human reader who wants to understand how things fit together.

The reason why I did this branch, is that I have a WIP branch which replaces the zone-names to identify zones with objects. The purpose of the branch is to move the knowledge/responsibility of zone-names/zone-identifiers from `Firewall.check_zone()` to `FirewallZone.get_zone()` -- which already needs to care about that.

Reply:
> delegate to FirewallZone for the validation.

Counter argument is that `fw_zone` shouldn't know nor care what the default zone is as that's determined by `fw`. As proved by the fact that you're calling `self._fw.get_default_zone()`.

So this improves one violation just to introduce another.

Generally all these `check_foo()` functions are useless IMO. I'd support removing them entirely.

Reply:
that wouldn't be true, if such separation would exist.

Instead, `fw_zone.py`'s `FirewallZone` has a `self._fw` attribute, which is used all over the place. Maybe that should be changed, but also, `self._fw` is `FirewallZone`'s access to the world. It's not clear, that it's really a problem to have or use it.

But yes, then all callers of `FirewallZone.get_zone()` and `FirewallZone.check_zone()` should instead resolve their zone-name name first. It does however mean, that a lot of obvious logic needs to move the layer up.

What the branch gives, is that `fw.py`'s `check_zone()` and `check_policy()` have no more logic at all (they should actually be dropped). They delegate everything to `self.zone.check_zone()` and `self.policy.get_policy()`. The logic of what a valid zone/policy is, is now at one place: in those related functions `get_zone()`, `check_zone()`, `get_zones()`, `get_policy()`, `get_policies()` (which all agree with each other and are defined size-by-side).